Triangulated MPCD geometry

jinnyjc opened the pull request:
## Description

<!-- Describe your changes in detail. -->

## Motivation and context

<!--- Why is this change required? What problem does it solve? -->

<!-- Replace ??? with the issue number that this pull request resolves. -->
Resolves #???

## How has this been tested?

<!--- Please describe in detail how you tested your changes. -->

<!--- Please build the sphinx documentation and check that any changes to
      documentation display properly. -->

## Checklist:

- [ ] I have reviewed the [**Contributor Guidelines**](https://github.com/glotzerlab/hoomd-blue/blob/trunk/CONTRIBUTING.rst).
- [ ] I agree with the terms of the [**HOOMD-blue Contributor Agreement**](https://github.com/glotzerlab/hoomd-blue/blob/trunk/ContributorAgreement.md).
- [ ] My name is on the list of contributors (`sphinx-doc/credits.rst`) in the pull request source branch.
- [ ] I have summarized these changes in `CHANGELOG.rst` following the established format.

jinnyjc committed:
Create TriangulatedGeometry data structure

jinnyjc committed:
Merge branch 'trunk' into feature/triangulated-geometry

jinnyjc committed:
Update TriangulatedGeometry data structure

jinnyjc commented:
@mphoward Could you please review what I have for `TriangulatedGeometry` when you get a chance?

mphoward commented on hoomd/mpcd/geometry.py:
```diff
@@ -490,6 +492,88 @@ def _attach_hook(self):
         super()._attach_hook()
 
 
+class TriangulatedGeometry(Geometry):
```
I think we may not want this object to inherit from `Geometry` because it doesn't follow the "is a" pattern, and a `TriangulatedGeometry` cannot be able to be used in places where a `Geometry` can.

You will want to make this inherit from `_HOOMDBaseObject` and copy the code to add the `no_slip` option.

mphoward commented on hoomd/mpcd/geometry.py:
```diff
@@ -490,6 +492,88 @@ def _attach_hook(self):
+        )
+        triangles = numpy.array([[0, 1, 3], [0, 2, 3]])
+
+        triangulatedgeometry = hoomd.mpcd.geometry.TriangulatedGeometry(
```
Do the vertices above correspond to a particular shape? For example, maybe you could make a tetrahedron. Then, call this `tetrahedron` so it's clear what this triangulated geometry is.

mphoward commented on hoomd/mpcd/geometry.py:
```diff
@@ -490,6 +492,88 @@ def _attach_hook(self):
+        "{inherited}", inspect.cleandoc(Geometry._doc_inherited)
+    )
+
+    def __init__(self, sysdef, vertices, triangles):
```
You won't need the system definition here.
```suggestion
    def __init__(self, vertices, triangles):
```

mphoward commented on hoomd/mpcd/geometry.py:
```diff
@@ -490,6 +492,88 @@ def _attach_hook(self):
+
+    def __init__(self, sysdef, vertices, triangles):
+        param_dict = ParameterDict(
+            sysdef=sysdef, vertices=vertices, triangles=triangles
```
Can you take a look for classes that accept array-like objects to see if we can do any validation on `vertices` and `triangles`?

mphoward commented on hoomd/mpcd/geometry.py:
```diff
@@ -490,6 +492,88 @@ def _attach_hook(self):
+        )
+        self._param_dict.update(param_dict)
+
+    def _attach_hook(self):
```
Take a look in other Python classes to see how to get the system definition. It should be available from some other member variable.

mphoward commented on hoomd/mpcd/geometry.py:
```diff
@@ -490,6 +492,88 @@ def _attach_hook(self):
+            self.sysdef, self.vertices, self.triangles
+        )
+        super()._attach_hook()
+
```
We will need to be careful here: should the `TriangulatedGeometry` retain all the vertices and triangles that are passed to the constructor as member variables? And, are these intended to be _global_ lists when you are running under MPI or not?

Regardless, this class should also have a property that returns the local accessor to the data (like the local snapshot has) so the user doesn't have to construct that accessor explicitly.

mphoward commented on hoomd/mpcd/geometry.py:
```diff
@@ -490,6 +492,88 @@ def _attach_hook(self):
+
+    def __init__(self, geometry):
+        self._geometry = geometry
+        self._simulation = geometry._simulation
```
Are these lines needed?

mphoward commented on hoomd/mpcd/geometry.py:
```diff
@@ -490,6 +492,88 @@ def _attach_hook(self):
+        return self
+
+    def __exit__(self, type, value, traceback):
+        self._cpp_obj.exit()
```
Does anything need to happen to clear `self._state`?

mphoward commented on hoomd/mpcd/geometry.py:
```diff
@@ -490,6 +492,88 @@ def _attach_hook(self):
+        return numpy.asarray(self._cpp_obj.getTriangles())
+
+
+class TriangulatedGeometryAccessHost(TriangulatedGeometryAccessBase):
```
Do you want to try adding the device accessor too?

mphoward commented on hoomd/mpcd/TriangulatedGeometry.cc:
```diff
@@ -0,0 +1,108 @@
+                unsigned int num_triangles = static_cast<unsigned int>(triangles.shape(0));
+
+                const Scalar3* v_ptr = reinterpret_cast<const Scalar3*>(vertices.data());
+                const uint3* t_ptr = reinterpret_cast<const uint3*>(triangles.data());
```
Can this be done by `static_cast` rather than `reinterpret_cast`?

mphoward reviewed (commented):
This overall looks like a really good draft, nice work! Can you take a look at these comments and let me know what questions you have?

jinnyjc committed:
Reupdate TriangulatedGeometry data structure

jinnyjc committed:
Implement streaming method for triangulated geometry

jinnyjc commented:
@mphoward I have implemented streaming method for triangulated geometry and very simple unit test for it. Could you please take a look when you get a chance?

mphoward commented on hoomd/mpcd/pytest/test_geometry.py:
```diff
@@ -256,3 +257,32 @@ def test_pickling(self, simulation_factory, snap):
+                [10, -10, -10],
+            ]
+        )
+        triangles = np.array([[0, 1, 3], [0, 2, 3], [4, 5, 7], [4, 6, 7]])
```
You probably want to force the data type to be safe
```suggestion
            ],
            dtype=float
        )
        triangles = np.array([[0, 1, 3], [0, 2, 3], [4, 5, 7], [4, 6, 7]], dtype=int)
```

mphoward commented on hoomd/mpcd/pytest/test_geometry.py:
```diff
@@ -256,3 +257,32 @@ def test_pickling(self, simulation_factory, snap):
+
+        geom = hoomd.mpcd.geometry.TriangulatedGeometry(sim, vertices, triangles, no_slip=True)
+
+        with geom.cpu_view as data:
```
This test looks clean to me. Can you check for the state how we would also test the `gpu_view` if numba is available?

mphoward commented on hoomd/mpcd/pytest/test_stream.py:
```diff
@@ -1359,3 +1359,106 @@ def test_check_mpcd_particles(self, simulation_factory, snap, R, expected_result
+                [5, -4, 5],
+            ]
+        )
+        triangles = np.array([[0, 1, 3], [0, 2, 3], [4, 5, 7], [4, 6, 7]])
```
```suggestion
            ],
            dtype=float
        )
        triangles = np.array([[0, 1, 3], [0, 2, 3], [4, 5, 7], [4, 6, 7]], dtype=int)
```

mphoward commented on hoomd/mpcd/pytest/test_stream.py:
```diff
@@ -1359,3 +1359,106 @@ def test_check_mpcd_particles(self, simulation_factory, snap, R, expected_result
+                snap.mpcd.velocity, [[-1.0, -1.0, 1.0]]
+            )
+
+    def test_step_slip(self, simulation_factory, snap):
```
This test looks good to me too! Is there a way to parametrize the slip vs. no sliip test to reduce duplication?

mphoward commented on hoomd/mpcd/TriangulatedGeometry.cc:
```diff
@@ -0,0 +1,127 @@
+      m_vertices(m_num_vertices, m_exec_conf), m_triangles(m_num_triangles, m_exec_conf),
+      m_no_slip(no_slip)
+    {
+    }
```
Do we need this constructor anymore? I like the second one better because it is more like the RAII pattern ("resource allocation is initialization").

mphoward reviewed (commented):
Thank you! This generally looks like a really good start to me. Could you please plan to walk me through the algorithm being implemented in the streaming method so I can be more targeted in my review? Some notes on the math sent via Slack would be especially helpful!

jinnyjc committed:
Add triangle unwrapping for periodic boundary

jinnyjc committed:
Update particle-triangle intersection

mphoward commented on hoomd/mpcd/geometry.py:
```diff
@@ -490,6 +492,109 @@ def _attach_hook(self):
+        "{inherited}", inspect.cleandoc(Geometry._doc_inherited)
+    )
+
+    def __init__(self, simulation, vertices, triangles, unwrap_distance, no_slip=True):
```
Is there a use case you had in mind where `unwrap_distance` needs to be different in each direction? A scalar will likely be more convenient most of the time.

mphoward commented on hoomd/mpcd/module.cc:
```diff
@@ -1,6 +1,7 @@
 // Copyright (c) 2009-2025 The Regents of the University of Michigan.
 // Part of HOOMD-blue, released under the BSD 3-Clause License.
 
+#include "TriangulatedGeometry.h"
```
I think you should not need to include this.
```suggestion
```

mphoward commented on hoomd/mpcd/stream.py:
```diff
@@ -349,9 +349,88 @@ def _register_cpp_class(cls, geometry, force, module, cpp_class_name):
+    This streaming method reflects MPCD particles from a triangulated surface
+    using specular reflections (bounce-back) rules using either "slip" or 
+    "no-slip" hydrodynamic boundary conditions. The external force is only applied
+    to the particles ath the beginning and the end of this process.
```
```suggestion
    to the particles at the beginning and the end of this process.
```

mphoward commented on hoomd/mpcd/TriangulatedGeometry.cc:
```diff
@@ -0,0 +1,210 @@
+    return m_unwrapped_triangles;
+    }
+
+void TriangulatedGeometry::unwrapTriangles()
```
If I understand it right, this unwrapping procedure is _very_ aggressive. It computes all periodic images of the triangles, not just the ones that are within `unwrap_distance` of the boundary.

You could use `Box::makeFraction` with the unwrap distance to figure out which these triangles have any of their vertices inside the unwrap distance, and only keep those.

I did have a question though: do you think it's important to keep both the original triangles _and_ the unwrapped ones (which will duplicate the original)? I imagined the data structure would be defined as the original triangles plus the unwrapped ones, all in one array.

If this unwrapping is CPU only for now (reasonable), to do this, I'd suggest using a `std::vector` to hold the unwrapped triangles that are within the unwrap distance and aren't in image `(0,0,0)` because you push back into a vector without knowing its size in advance. Then, I would resize your existing triangle data structure and copy those unwrapped ones into it. That will allow you to only keep 1 triangle data structure for traversal, rather than worry about two separate arrays.

mphoward commented on hoomd/mpcd/TriangulatedGeometry.h:
```diff
@@ -0,0 +1,147 @@
+    unsigned int getNumUnwrappedTriangles() const;
+
+    const GPUArray<Scalar3>& getUnwrappedVertices() const;
+    const GPUArray<uint3>& getUnwrappedTriangles() const;
```
See longer comment above. I would think you would want to remove these, unless you have a use case in mind where we want both copies.

mphoward commented on hoomd/mpcd/TriangulatedGeometry.h:
```diff
@@ -0,0 +1,147 @@
+    bool getNoSlip() const;
+
+    private:
+    void unwrapTriangles();
```
I prefer to put the methods after the member variables within each section of visibility. Please also make sure to start adding doxygen-style docstrings to variables and methods!

mphoward commented on hoomd/mpcd/TriangulatedGeometryStreamingMethod.h:
```diff
@@ -0,0 +1,352 @@
+
+    {
+    // calculate dimension where the ray direction is maximal
+    const Scalar ax = fabs(vel_v.x);
```
Check in `HOOMDMath.h` for `fabs` in the fast (preferred) or slow namespace. I think it is there, and you want to make sure you use that for when this goes to GPU code.

mphoward commented on hoomd/mpcd/TriangulatedGeometryStreamingMethod.h:
```diff
@@ -0,0 +1,352 @@
+    {
+namespace mpcd
+    {
+//! MPCD confined streaming method
```
This documentation belongs with the class, make sure it goes before it.

mphoward commented on hoomd/mpcd/TriangulatedGeometryStreamingMethod.h:
```diff
@@ -0,0 +1,352 @@
+ *
+ */
+
+inline bool intersectTriangle(const vec3<Scalar>& pos_v,
```
Make sure to add some documentation including the reference to the paper you are following.

Also, do these need to be `vec3<Scalar>` or would `Scalar3` work? You are probably storing the relevant data in `Scalar3`, so using it directly avoids one cast. (Probably trivial in the scheme of things, but still better to skip if not necessary).

mphoward commented on hoomd/mpcd/TriangulatedGeometryStreamingMethod.h:
```diff
@@ -0,0 +1,352 @@
+
+                // degenerate triangles
+                if (nn <= Scalar(0))
+                    continue;
```
Is this possible in typical use? We could sanity check the geometry when it is created, for example, so we don't have to do this check all the time later.

If this check isn't needed here, I think this stuff can probably all be moved to the case where `found = true` so you only compute it once for the actual collision because at this stage, all you care about is the hit time, right?

mphoward commented on hoomd/mpcd/TriangulatedGeometryStreamingMethod.h:
```diff
@@ -0,0 +1,352 @@
+                best_pos = pos_v + vel_v * t_hit;
+                }
+
+            if (!found)
```
Suggest to flip these branches to `if (found)` since it's the more complicated one, and default to simple streaming otherwise.

mphoward commented on hoomd/mpcd/TriangulatedGeometryStreamingMethod.h:
```diff
@@ -0,0 +1,352 @@
+    } // end namespace detail
+    } // end namespace mpcd
+    } // end namespace hoomd
+#endif // MPCD_TRIANGULATED_GEOMETRY_STREAMING_METHOD_H_
```
GitHub is marking this is missing an EOF. That would normally be taken care of by prek / pre-commit, so make sure you have that configured.

mphoward reviewed (commented):
This is starting to shape up! Here are some comments, please let me know what questions you have.

jinnyjc committed:
Address review comments

jinnyjc committed:
Fix infinite loop in collision handling

jinnyjc commented:
@mphoward Could you please review my changes? Here's a list of the changes:

- Added cross product in `HOOMDMath.h` to support operations using `Scalar3` instead of `vec3<Scalar>`.
- Remove `#include "TriangulatedGeometry.h"` from `module.cc`.
- Changed `unwrap_distance` from `Scalar3` to `Scalar` to use a uniform unwrap distance in all directions.
- Updated the unwrapping algorithm to selectively unwrap triangles within a buffer region instead of duplicating all periodic images.
- Merged two arrays (original and unwrapped) into a single array. Added access methods to distinguish between original geometry and combined (original + unwrapped) geometry.
- Updated collision detection so that collision time is constrained between 0 and `dt_remaining`. Collisions are ignored when the velocity and triangle normal do not point in the same direction. This update requires triangle ordering to ensure normals point outward from the fluid region.

jinnyjc committed:
Add triangle BVH for triangulated geometry

jinnyjc committed:
Implement GPU triangulated geometry streaming method

jinnyjc commented:
@mphoward Could you please review my GPU triangulated geometry streaming method?

jinnyjc committed:
Add max_bounce to triangulated geometry streaming

jinnyjc committed:
Merge remote-tracking branch 'upstream/trunk' into feature/triangulated-geometry

mphoward commented:
@jinnyjc please remove the stale label!
